File: swift/BlackjackTrainer/BlackjackTrainer/Models/StrategyChart.swift
import Foundation



// MARK: - Strategy Chart Implementation

struct StrategyChart: Codable {
    private let hardTotals: [String: Action]
    private let softTotals: [String: Action]
    private let pairs: [String: Action]
    private let mnemonics: [String: String]
    
    init() {
        self.hardTotals = Self.buildHardTotals()
        self.softTotals = Self.buildSoftTotals()
        self.pairs = Self.buildPairs()
        self.mnemonics = Self.buildMnemonics()
    }
    
    // MARK: - Strategy Chart Data Construction
    
    private static func buildHardTotals() -> [String: Action] {
        var chart: [String: Action] = [:]
        
        // Hard 5-8: Always hit
        for total in 5...8 {
            for dealer in 2...11 {
                chart["\(total)-\(dealer)"] = .hit
            }
        }
        
        // Hard 9: Double vs 3-6, otherwise hit
        for dealer in 2...11 {
            let action: Action = (3...6).contains(dealer) ? .double : .hit
            chart["9-\(dealer)"] = action
        }
        
        // Hard 10: Double vs 2-9, otherwise hit
        for dealer in 2...11 {
            let action: Action = (2...9).contains(dealer) ? .double : .hit
            chart["10-\(dealer)"] = action
        }
        
        // Hard 11: Double vs 2-10, hit vs Ace
        for dealer in 2...11 {
            let action: Action = dealer <= 10 ? .double : .hit
            chart["11-\(dealer)"] = action
        }
        
        // Hard 12: Stand vs 4-6, otherwise hit
        for dealer in 2...11 {
            let action: Action = (4...6).contains(dealer) ? .stand : .hit
            chart["12-\(dealer)"] = action
        }
        
        // Hard 13-16: Stand vs 2-6, otherwise hit
        for total in 13...16 {
            for dealer in 2...11 {
                let action: Action = (2...6).contains(dealer) ? .stand : .hit
                chart["\(total)-\(dealer)"] = action
            }
        }
        
        // Hard 17+: Always stand
        for total in 17...21 {
            for dealer in 2...11 {
                chart["\(total)-\(dealer)"] = .stand
            }
        }
        
        return chart
    }
    
    private static func buildSoftTotals() -> [String: Action] {
        var chart: [String: Action] = [:]
        
        // Soft 13-14 (A,2-A,3): Double vs 5-6, otherwise hit
        for total in [13, 14] {
            for dealer in 2...11 {
                let action: Action = (5...6).contains(dealer) ? .double : .hit
                chart["\(total)-\(dealer)"] = action
            }
        }
        
        // Soft 15-16 (A,4-A,5): Double vs 4-6, otherwise hit
        for total in [15, 16] {
            for dealer in 2...11 {
                let action: Action = (4...6).contains(dealer) ? .double : .hit
                chart["\(total)-\(dealer)"] = action
            }
        }
        
        // Soft 17 (A,6): Double vs 3-6, otherwise hit
        for dealer in 2...11 {
            let action: Action = (3...6).contains(dealer) ? .double : .hit
            chart["17-\(dealer)"] = action
        }
        
        // Soft 18 (A,7): Stand vs 2,7,8; Double vs 3-6; Hit vs 9,10,A
        for dealer in 2...11 {
            let action: Action
            if [2, 7, 8].contains(dealer) {
                action = .stand
            } else if (3...6).contains(dealer) {
                action = .double
            } else {
                action = .hit
            }
            chart["18-\(dealer)"] = action
        }
        
        // Soft 19-21: Always stand
        for total in 19...21 {
            for dealer in 2...11 {
                chart["\(total)-\(dealer)"] = .stand
            }
        }
        
        return chart
    }
    
    private static func buildPairs() -> [String: Action] {
        var chart: [String: Action] = [:]
        
        // A,A: Always split
        for dealer in 2...11 {
            chart["11-\(dealer)"] = .split
        }
        
        // 2,2 and 3,3: Split vs 2-7, otherwise hit
        for pairValue in [2, 3] {
            for dealer in 2...11 {
                let action: Action = (2...7).contains(dealer) ? .split : .hit
                chart["\(pairValue)-\(dealer)"] = action
            }
        }
        
        // 4,4: Split vs 5-6, otherwise hit
        for dealer in 2...11 {
            let action: Action = (5...6).contains(dealer) ? .split : .hit
            chart["4-\(dealer)"] = action
        }
        
        // 5,5: Never split, treat as hard 10
        for dealer in 2...11 {
            let action: Action = (2...9).contains(dealer) ? .double : .hit
            chart["5-\(dealer)"] = action
        }
        
        // 6,6: Split vs 2-6, otherwise hit
        for dealer in 2...11 {
            let action: Action = (2...6).contains(dealer) ? .split : .hit
            chart["6-\(dealer)"] = action
        }
        
        // 7,7: Split vs 2-7, otherwise hit
        for dealer in 2...11 {
            let action: Action = (2...7).contains(dealer) ? .split : .hit
            chart["7-\(dealer)"] = action
        }
        
        // 8,8: Always split
        for dealer in 2...11 {
            chart["8-\(dealer)"] = .split
        }
        
        // 9,9: Split vs 2-9 except 7, stand vs 7,10,A
        for dealer in 2...11 {
            let action: Action
            if [7, 10, 11].contains(dealer) {
                action = .stand
            } else {
                action = .split
            }
            chart["9-\(dealer)"] = action
        }
        
        // 10,10: Never split, always stand
        for dealer in 2...11 {
            chart["10-\(dealer)"] = .stand
        }
        
        return chart
    }
    
    private static func buildMnemonics() -> [String: String] {
        return [
            "dealer_weak": "Dealer bust cards (4,5,6) = player gets greedy",
            "always_split": "Aces and eights, don't hesitate",
            "never_split": "Tens and fives, keep them alive",
            "teens_vs_strong": "Teens stay vs weak, flee from strong",
            "soft_17": "A,7 is the tricky soft hand",
            "hard_12": "12 is the exception - only stand vs 4,5,6",
            "doubles": "Double when dealer is weak and you can improve"
        ]
    }
    
    // MARK: - Strategy Chart Interface
    
    func getCorrectAction(for scenario: GameScenario) -> Action {
        let key = "\(scenario.playerTotal)-\(scenario.dealerCard.value)"
        
        switch scenario.handType {
        case .hard:
            return hardTotals[key] ?? .hit
        case .soft:
            return softTotals[key] ?? .hit
        case .pair:
            return pairs[key] ?? .hit
        }
    }
    
    func getExplanation(for scenario: GameScenario) -> String {
        // Check for specific scenario explanations
        let specificExplanations: [(HandType, Int) -> String?] = [
            { handType, playerTotal in
                if handType == .pair && playerTotal == 11 { return mnemonics["always_split"] }
                return nil
            },
            { handType, playerTotal in
                if handType == .pair && playerTotal == 8 { return mnemonics["always_split"] }
                return nil
            },
            { handType, playerTotal in
                if handType == .pair && [10, 5].contains(playerTotal) { return mnemonics["never_split"] }
                return nil
            },
            { handType, playerTotal in
                if handType == .soft && playerTotal == 18 { return mnemonics["soft_17"] }
                return nil
            },
            { handType, playerTotal in
                if handType == .hard && playerTotal == 12 { return mnemonics["hard_12"] }
                return nil
            }
        ]
        
        for explanation in specificExplanations {
            if let result = explanation(scenario.handType, scenario.playerTotal) {
                return result
            }
        }
        
        // Check dealer strength patterns
        let dealerStrength = DealerStrength.from(card: scenario.dealerCard)
        if dealerStrength == .weak {
            return mnemonics["dealer_weak"] ?? "Follow basic strategy patterns"
        }
        
        if scenario.handType == .hard && (13...16).contains(scenario.playerTotal) && dealerStrength == .strong {
            return mnemonics["teens_vs_strong"] ?? "Follow basic strategy patterns"
        }
        
        return "Follow basic strategy patterns"
    }
    
    func isAbsoluteRule(for scenario: GameScenario) -> Bool {
        let absolutes: [(HandType, Int)] = [
            (.pair, 11),  // Always split A,A
            (.pair, 8),   // Always split 8,8
            (.pair, 10),  // Never split 10,10
            (.pair, 5),   // Never split 5,5
        ]
        
        // Check for specific absolute rules
        if absolutes.contains(where: { $0.0 == scenario.handType && $0.1 == scenario.playerTotal }) {
            return true
        }
        
        // Hard 17+ always stand
        if scenario.handType == .hard && scenario.playerTotal >= 17 {
            return true
        }
        
        // Soft 19+ always stand
        if scenario.handType == .soft && scenario.playerTotal >= 19 {
            return true
        }
        
        return false
    }
}

File: android/app/src/main/kotlin/net/kristopherjohnson/blackjacktrainer/data/model/StrategyChart.kt
package net.kristopherjohnson.blackjacktrainer.data.model

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import net.kristopherjohnson.blackjacktrainer.domain.model.*
import timber.log.Timber

/**
 * Complete basic strategy chart implementation
 * Based on blackjack_basic_strategy.md specification
 */
data class StrategyChart(
    val hardTotals: Map<HandKey, Action>,
    val softTotals: Map<HandKey, Action>,
    val pairs: Map<HandKey, Action>,
    val mnemonics: Map<String, String>,
    val dealerGroups: Map<DealerStrength, List<Int>>,
    val version: Int = 1
) {
    
    /**
     * Get the correct action for a given scenario
     */
    suspend fun getCorrectAction(scenario: GameScenario): Result<Action> = withContext(Dispatchers.Default) {
        try {
            val key = HandKey(scenario.playerTotal, scenario.dealerCard.value)
            val action = when (scenario.handType) {
                HandType.HARD -> hardTotals[key]
                HandType.SOFT -> softTotals[key]
                HandType.PAIR -> pairs[key]
            }
            
            action?.let { Result.success(it) } 
                ?: Result.failure(
                    StrategyException.InvalidScenario(
                        "No strategy found for ${scenario.handType} ${scenario.playerTotal} vs ${scenario.dealerCard.displayValue}",
                        scenario.copy()
                    )
                )
        } catch (e: Exception) {
            Timber.e(e, "Strategy lookup failed for scenario: $scenario")
            Result.failure(StrategyException.DataCorruption("Strategy data corrupted", e))
        }
    }
    
    /**
     * Get explanation/mnemonic for a scenario
     */
    fun getExplanation(scenario: GameScenario): String {
        val key = "${scenario.handType}-${scenario.playerTotal}-${scenario.dealerCard.value}"
        return mnemonics[key] ?: getGeneralExplanation(scenario)
    }
    
    private fun getGeneralExplanation(scenario: GameScenario): String {
        return when (scenario.handType) {
            HandType.HARD -> when {
                scenario.playerTotal <= 8 -> "Always hit low hard totals"
                scenario.playerTotal >= 17 -> "Always stand on hard 17+"
                scenario.dealerCard.value in 4..6 -> "Dealer has bust card - be conservative"
                scenario.dealerCard.value >= 9 -> "Dealer is strong - hit to improve"
                else -> "Follow basic strategy patterns"
            }
            HandType.SOFT -> when {
                scenario.playerTotal >= 19 -> "Always stand on soft 19+"
                scenario.dealerCard.value in 4..6 -> "Double against dealer bust cards"
                else -> "Soft hands are flexible - be aggressive"
            }
            HandType.PAIR -> when {
                scenario.playerTotal == 22 -> "Always split Aces"
                scenario.playerTotal == 16 && scenario.playerCards.first().rank == Rank.EIGHT -> "Always split 8s"
                scenario.playerTotal == 20 -> "Never split 10s"
                else -> "Consider dealer strength for split decisions"
            }
        }
    }
    
    companion object {
        fun createDefault(): StrategyChart {
            return StrategyChart(
                hardTotals = initializeHardTotals(),
                softTotals = initializeSoftTotals(),
                pairs = initializePairs(),
                mnemonics = initializeMnemonics(),
                dealerGroups = mapOf(
                    DealerStrength.WEAK to listOf(4, 5, 6),
                    DealerStrength.MEDIUM to listOf(2, 3, 7, 8),
                    DealerStrength.STRONG to listOf(9, 10, 11)
                )
            )
        }
        
        private fun initializeHardTotals(): Map<HandKey, Action> {
            val strategy = mutableMapOf<HandKey, Action>()
            
            // 8 or less - Always Hit
            for (total in 5..8) {
                for (dealer in 2..11) {
                    strategy[HandKey(total, dealer)] = Action.HIT
                }
            }
            
            // 9 - Hit except double vs 3-6
            for (dealer in 2..11) {
                strategy[HandKey(9, dealer)] = when (dealer) {
                    3, 4, 5, 6 -> Action.DOUBLE
                    else -> Action.HIT
                }
            }
            
            // 10 - Double vs 2-9, Hit vs 10,A
            for (dealer in 2..11) {
                strategy[HandKey(10, dealer)] = when (dealer) {
                    10, 11 -> Action.HIT
                    else -> Action.DOUBLE
                }
            }
            
            // 11 - Double vs 2-10, Hit vs A
            for (dealer in 2..11) {
                strategy[HandKey(11, dealer)] = when (dealer) {
                    11 -> Action.HIT
                    else -> Action.DOUBLE
                }
            }
            
            // 12 - Stand vs 4-6, Hit vs rest
            for (dealer in 2..11) {
                strategy[HandKey(12, dealer)] = when (dealer) {
                    4, 5, 6 -> Action.STAND
                    else -> Action.HIT
                }
            }
            
            // 13-16 - Stand vs 2-6, Hit vs 7-A
            for (total in 13..16) {
                for (dealer in 2..11) {
                    strategy[HandKey(total, dealer)] = when (dealer) {
                        2, 3, 4, 5, 6 -> Action.STAND
                        else -> Action.HIT
                    }
                }
            }
            
            // 17+ - Always Stand
            for (total in 17..21) {
                for (dealer in 2..11) {
                    strategy[HandKey(total, dealer)] = Action.STAND
                }
            }
            
            return strategy
        }
        
        private fun initializeSoftTotals(): Map<HandKey, Action> {
            val strategy = mutableMapOf<HandKey, Action>()
            
            // A,2 (13) - Hit except Double vs 5,6
            for (dealer in 2..11) {
                strategy[HandKey(13, dealer)] = when (dealer) {
                    5, 6 -> Action.DOUBLE
                    else -> Action.HIT
                }
            }
            
            // A,3 (14) - Hit except Double vs 5,6
            for (dealer in 2..11) {
                strategy[HandKey(14, dealer)] = when (dealer) {
                    5, 6 -> Action.DOUBLE
                    else -> Action.HIT
                }
            }
            
            // A,4 (15) - Hit except Double vs 4,5,6
            for (dealer in 2..11) {
                strategy[HandKey(15, dealer)] = when (dealer) {
                    4, 5, 6 -> Action.DOUBLE
                    else -> Action.HIT
                }
            }
            
            // A,5 (16) - Hit except Double vs 4,5,6
            for (dealer in 2..11) {
                strategy[HandKey(16, dealer)] = when (dealer) {
                    4, 5, 6 -> Action.DOUBLE
                    else -> Action.HIT
                }
            }
            
            // A,6 (17) - Hit except Double vs 3,4,5,6
            for (dealer in 2..11) {
                strategy[HandKey(17, dealer)] = when (dealer) {
                    3, 4, 5, 6 -> Action.DOUBLE
                    else -> Action.HIT
                }
            }
            
            // A,7 (18) - Stand vs 2,7,8; Double vs 3,4,5,6; Hit vs 9,10,A
            for (dealer in 2..11) {
                strategy[HandKey(18, dealer)] = when (dealer) {
                    2, 7, 8 -> Action.STAND
                    3, 4, 5, 6 -> Action.DOUBLE
                    9, 10, 11 -> Action.HIT
                    else -> Action.STAND
                }
            }
            
            // A,8 (19) - Always Stand
            for (dealer in 2..11) {
                strategy[HandKey(19, dealer)] = Action.STAND
            }
            
            // A,9 (20) - Always Stand
            for (dealer in 2..11) {
                strategy[HandKey(20, dealer)] = Action.STAND
            }
            
            return strategy
        }
        
        private fun initializePairs(): Map<HandKey, Action> {
            val strategy = mutableMapOf<HandKey, Action>()
            
            // A,A - Always Split
            for (dealer in 2..11) {
                strategy[HandKey(22, dealer)] = Action.SPLIT
            }
            
            // 10,10 - Never Split
            for (dealer in 2..11) {
                strategy[HandKey(20, dealer)] = Action.NO_SPLIT
            }
            
            // 9,9 - Split vs 2-6,8,9; Don't split vs 7,10,A
            for (dealer in 2..11) {
                strategy[HandKey(18, dealer)] = when (dealer) {
                    7, 10, 11 -> Action.NO_SPLIT
                    else -> Action.SPLIT
                }
            }
            
            // 8,8 - Always Split
            for (dealer in 2..11) {
                strategy[HandKey(16, dealer)] = Action.SPLIT
            }
            
            // 7,7 - Split vs 2-7; Don't split vs 8-A
            for (dealer in 2..11) {
                strategy[HandKey(14, dealer)] = when (dealer) {
                    2, 3, 4, 5, 6, 7 -> Action.SPLIT
                    else -> Action.NO_SPLIT
                }
            }
            
            // 6,6 - Split vs 2-6; Don't split vs 7-A
            for (dealer in 2..11) {
                strategy[HandKey(12, dealer)] = when (dealer) {
                    2, 3, 4, 5, 6 -> Action.SPLIT
                    else -> Action.NO_SPLIT
                }
            }
            
            // 5,5 - Never Split
            for (dealer in 2..11) {
                strategy[HandKey(10, dealer)] = Action.NO_SPLIT
            }
            
            // 4,4 - Don't split except vs 5,6
            for (dealer in 2..11) {
                strategy[HandKey(8, dealer)] = when (dealer) {
                    5, 6 -> Action.SPLIT
                    else -> Action.NO_SPLIT
                }
            }
            
            // 3,3 - Split vs 2-7; Don't split vs 8-A
            for (dealer in 2..11) {
                strategy[HandKey(6, dealer)] = when (dealer) {
                    2, 3, 4, 5, 6, 7 -> Action.SPLIT
                    else -> Action.NO_SPLIT
                }
            }
            
            // 2,2 - Split vs 2-7; Don't split vs 8-A
            for (dealer in 2..11) {
                strategy[HandKey(4, dealer)] = when (dealer) {
                    2, 3, 4, 5, 6, 7 -> Action.SPLIT
                    else -> Action.NO_SPLIT
                }
            }
            
            return strategy
        }
        
        private fun initializeMnemonics(): Map<String, String> {
            return mapOf(
                // Hard totals
                "HARD-9-3" to "Double 9 vs dealer weak cards (3-6)",
                "HARD-10-10" to "Don't double 10 vs dealer 10 or Ace",
                "HARD-11-11" to "Don't double 11 vs dealer Ace",
                "HARD-12-4" to "Stand 12 vs dealer bust cards (4-6)",
                "HARD-16-7" to "Hit 16 vs dealer strong cards (7+)",
                
                // Soft totals
                "SOFT-18-9" to "A,7 is tricky - hit vs strong dealer cards",
                "SOFT-17-3" to "Double soft 17 vs dealer weak cards",
                
                // Pairs
                "PAIR-22-2" to "Aces and eights, don't hesitate",
                "PAIR-16-2" to "Always split 8s - turn disaster into opportunity",
                "PAIR-20-2" to "Never split 10s - 20 is too good to break up",
                "PAIR-18-7" to "Don't split 9s vs 7 - 18 vs 7 is good",
                
                // General patterns
                "dealer_weak" to "Dealer bust cards (4,5,6) = player gets greedy",
                "dealer_strong" to "Dealer strong (9,10,A) = player plays conservatively",
                "always_split" to "Always split Aces and 8s",
                "never_split" to "Never split 10s and 5s",
                "teens_vs_strong" to "Teen totals (13-16) flee from strong dealer cards"
            )
        }
    }
}

/**
 * Key for strategy lookup
 */
data class HandKey(
    val playerTotal: Int,
    val dealerCard: Int
)

/**
 * Strategy-related exceptions
 */
sealed class StrategyException(message: String, cause: Throwable? = null) : Exception(message, cause) {
    data class InvalidScenario(val details: String, val scenario: GameScenario) : StrategyException(details)
    data class DataCorruption(val details: String, override val cause: Throwable) : StrategyException(details, cause)
}